### PR TITLE
Reduces heap allocations for the some byte[] uses

### PR DESCRIPTION
**Pull request by jacobslusser**
Profiling reveals that creation of `byte[4]` arrays are our most common heap objects. Many of these come from reading integers from the byte stream where we allocate a small `byte[4]` or `byte[8]` array and then convert it to a 32 or 64-bit integer. 

Results from a simple test of using the `SftpClient` to upload and download some files:

**Before: when byte[] arrays were created for each integer read**
Screenshot 2023-12-10 181955

**After: when byte[] are allocated and read from the stack**
Screenshot 2023-12-10 182136

**Changes**
- Creates a new `SshDataStream.ReadBytes` overload that supports `Span<byte>` for .NET Standard 2.1+ and .NET 6+ targets.
- Updates `SshDataStream.ReadUInt32` and `SshDataStream.ReadUInt64` to read onto the stack and use newer `BinaryPrimitives` class for integer conversion for .NET Standard 2.1+ and .NET 6+ targets.
- Updates `SshData.ReadUInt32` and `SshData.ReadUInt64` to use the existing stream methods to avoid another temporary buffer.

### CODE REVIEW

**Comment by Rob-Hague** [comment]
Funny, I had very similar changes in #1138 🙂 Feel free to cherry-pick some of those, or otherwise this looks good to me

File: src/Renci.SshNet/Common/SshDataStream.cs
Comment (WojciechNagorski) [inline src/Renci.SshNet/Common/SshDataStream.cs:291]: Wouldn't it be better to throw exceptions from separate methods? I don't remember exactly, but it was probably better for performance reasons. Throwing exceptions probably blocked the method inlining.

This doesn't block this PR, it's more of a question.

Comment (Rob-Hague) [reply src/Renci.SshNet/Common/SshDataStream.cs:291]: The benefit of using "throw helpers" is that the code size of the method is reduced by shifting the "cold" throwing code into another method (which is not inlined).

See [this example](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdACJQQDmMA9iglAMYpUDCzAJgFMAgjAgAbAJ4ooKANxFSFagHFxzYBKgAvCO2YwFxcpSoBJAPJHMAZgrYyPMiDIBZQWGYAnSQGUEXoIQYGQA3kRkkRR2mAAsZABKQfwAQpIIgii4ABS+AA4QMAA8wOmCAHxkwACuAGa1gl4AlGERUe0AbhBeVWUoSRD8ZAC8icnZNfWNTUbt7VC1ZBN9A0NFVXUNXlQAMoIwDAgAFk1tc62E5+fHXswA7mQwgg/CXgzVYPsIFtXftQmFBiCACiAA8OII8voYNkxJ9mLUJptpnBlCQqAAxbyQBDZHjVcQIaqBMwwWrMcwwLpeegwBD4wnEwSogBEABUjoIyIEAI7VTIZIbifaHI5LUIkAC+LVkZAYgT0jTIx0Kys5ZAgHCJEkeH2ASoRvQyKG5yXFuGlVBZqMmW12IuONpWySaMzOc0l7s9lyi7tsFHiqzSxuwuQKxVKGUqtumFyuZGpRsyqxGY0GSKmzVmVwWS0jybN6xj2z2B2Opx953ClfjZA5tzuGa2TuNqzdNfa3vj7vaAG0aMxMgA5ZgIJJErwwAC6PailAAbAG60cG2HCiUytHkV5UbAEEn+i7Z5Fq7Wojd7o9nmRXu9PnSfn8AQcQeDIdDYcFBAimyi0ZjsT0PECQnQRSXJSlqVpekQKZVkOS5Xl+TYQQhQdMVsglaUyFleUggyHoVRgNUuU1bVxF1MB9R6Q18xNBUhkwi0mitG1t3tMsjhbAtBldbNzklbsa29SUgA). The generated code for `ReadBytes2` is much smaller than `ReadBytes1`.

In practice, I don't think we need to be thinking at this level of optimisation just yet

**Review by WojciechNagorski — APPROVED** [review]

**Comment by jacobslusser** [comment]
> Funny, I had very similar changes in #1138 🙂 Feel free to cherry-pick some of those, or otherwise this looks good to me

@Rob-Hague wow, your changes are more than a little similar. At one point I had even ported the `ReadExactly` just as you had but decided instead to follow our existing pattern. I will definitely look at merging the others. 

Why didn't you ever merge this?

**Comment by Rob-Hague** [comment]
I think I wanted to justify it with some profiling like you did, but got stuck into some other work

**Comment by WojciechNagorski** [comment]
I will merge any optimization if it has justify as goof as this PR 😃. https://github.com/sshnet/SSH.NET/pull/1138 was a draft.

**Comment by WojciechNagorski** [comment]
The 2023.0.1 version has been released to Nuget: https://www.nuget.org/packages/SSH.NET/2023.0.1

**Pull request merged**
